style: how to split functions (#2075)
tigerbeetle/tigerbeetle: docs/TIGER_STYLE.md

- Restrict the length of function bodies to reduce the probability of poorly structured code. We
  enforce a **hard limit of 70 lines per function**.

>   Splitting code into functions requires taste: there are may ways to cut a wall of code into into
>   chunks of 70 lines, but only few splits will feel right. Some rules of thumb here:

Reviewer: ```suggestion
  Splitting code into functions requires taste. There are many ways to cut a wall of code into
  chunks of 70 lines, but only a few splits will feel right. Some rules of thumb:
```

>   * Good function's shape is an inverse that of an hourglass — few parameters, a simple return
>     type, but a lot of meaty logic between the braces.

Reviewer: ```suggestion
  * Good function shape is often the inverse of an hourglass: a few parameters, a simple return
    type, and a lot of meaty logic between the braces.
```

>   * Centralize control flow: when splitting a large function, try to keep all switch an if
>     statements in the "parent" function, and move non-branchy fragments of logic to helper
>     functions. Divide responsibility: all control flow is handled by _one_ function, the rest don't
>     care about control flow at all.

Reviewer: ```suggestion
  * Centralize control flow. When splitting a large function, try to keep all switch/if
    statements in the "parent" function, and move non-branchy logic fragments to helper
    functions. Divide responsibility. All control flow should be handled by _one_ function, the rest shouldn't
    care about control flow at all. In other words,
    ["push ifs up and fors down"](https://matklad.github.io/2023/11/15/push-ifs-up-and-fors-down.html).
```

>     [Push ifs up](https://matklad.github.io/2023/11/15/push-ifs-up-and-fors-down.html).

Reviewer: ```suggestion
```

>   * Similarly, centralize state manipulation. Let the parent function keep all the relevant state in
>     local variables, and use helpers to compute what needs to change, rather than applying the
>     change directly. Keep leaf functions pure.

Reviewer: ```suggestion
  * Similarly, centralize state manipulation. Let the parent function keep all relevant state in
    local variables, and use helpers to compute what needs to change, rather than applying the
    change directly. Keep leaf functions pure.
```
